test/unit/carriers/correios_test.rb

```diff
@@ -0,0 +1,160 @@
+    options = { 
+      :company_id => 1010,
+      :password => 123123,
+      :declared_value_extra => 10.50,
+      :delivery_notice_extra => true,
+      :mao_propria_extra => true
+    }
+
+    @carrier.expects(:perform).with([url]).returns([@response_book_success])
+    @carrier.find_rates(@saopaulo, @patosdeminas, [@book], options)
+  end
+
+  def test_book_response
+    @carrier.stubs(:perform).returns([@response_book_success])
+    response = @carrier.find_rates(@saopaulo, @patosdeminas, [@book])
+
+    assert_equal 1, response.rates.size
+    assert_equal [10520], response.rates.map(&:price)
+  end
+
+  def test_poster_response
+    @carrier.stubs(:perform).returns([@response_poster_success])
+    response = @carrier.find_rates(@saopaulo, @patosdeminas, [@poster])
+
+    assert_equal 2, response.rates.size
+    assert_equal [1000, 2000], response.rates.map(&:price)
+  end
+
+  def test_two_books_response
+    @carrier.stubs(:perform).returns([@response_book_success, @response_book_success])
+    response = @carrier.find_rates(@saopaulo, @patosdeminas, [@book, @book])
+
+    assert_equal 1, response.rates.size
+    assert_equal [21040], response.rates.map(&:price)
+  end
+
+  def test_two_posters_response
+    @carrier.stubs(:perform).returns([@response_poster_success, @response_poster_success])
+    response = @carrier.find_rates(@saopaulo, @patosdeminas, [@poster, @poster])
+
+    assert_equal 2, response.rates.size
+    assert_equal [2000, 4000], response.rates.map(&:price)
+  end
+
+  def test_response_parsing
+    @carrier.stubs(:perform).returns([@response_clothes, @response_shoes])
+    response = @carrier.find_rates(@saopaulo, @patosdeminas, [@book, @book])
+    service_codes = [41106, 41300, 40215, 81019]
+    service_names = [
+      'PAC sem contrato',
+      'PAC para grandes formatos',
+      'SEDEX 10, sem contrato',
+      'e-SEDEX, com contrato'
+    ]
+
+    assert_equal service_codes, response.rates.map(&:service_code)
+    assert_equal service_names, response.rates.map(&:service_name)
+  end
+
+  def test_response_params_options
+    @carrier.stubs(:perform).returns([@response_book_success])
+    response = @carrier.find_rates(@saopaulo, @patosdeminas, [@book])
+
+    assert_equal [Nokogiri::XML::Document], response.params['responses'].map(&:class)
+    assert_equal [Nokogiri::XML(@response_book_success).to_xml], response.params['responses'].map(&:to_xml)
+  end
+
+  def test_book_invalid_response
+    @carrier.stubs(:perform).returns([@response_book_invalid])
+    @carrier.find_rates(@saopaulo, @patosdeminas, [@book])
+
+    refute true, 'Must raise ActiveShipping::ResponseError'
+  rescue => error
+    assert_kind_of ActiveShipping::ResponseError, error
+    assert_equal "CEP de origem invalido", error.message
+    assert_equal error.response.raw_responses, [@response_book_invalid]
+    assert_equal Hash.new, error.response.params
+  end
```

Comment:
Also, can you add a test where for a `find_rates` call with two packages, in which the first request succeeds and the second requests fails? I think the best way to handle that is to return an error response in that case.